HeinOnline.js

```diff
@@ -0,0 +1,240 @@
+		lines[i] = lines[i].replace(/^YR  /, "DA  ");
+		lines[i] = lines[i].replace(/^OP  /, "EP  ");
+	}
+	ris = lines.join("\n");
+	return ris;
+}
+
+// Not all pages have a downloadable PDF
+function translateRIS(ris, pdfURL) {
+	var trans = Zotero.loadTranslator('import');
+	trans.setTranslator('32d59d2d-b65a-4da4-b0a3-bdd3cfb979e7');//https://github.com/zotero/translators/blob/master/RIS.js
+	trans.setString(ris);
+	if (pdfURL) {
+		trans.setHandler('itemDone', function (obj, item) {
+			item.attachments = [{
+				title: "Full Text PDF",
+				url: pdfURL,
+				mimeType: "application/pdf"
+			}];
+			item.complete();
+		});
+	} else {
+		trans.setHandler('itemDone', function (obj, item) {
+			item.complete();
+		});
+	}
+	trans.translate();
+}
+
+function translateCOinS(COinS) {
+    var item = new Zotero.Item();
+    try {
+        Zotero.Utilities.parseContextObject(COinS, item);
+        item.complete();
+    } catch(e) {
+        Zotero.debug("error parsing COinS");
+    }
+}
+
+// Build URL for RIS, and for PDF if available
+function scrapePage(doc, url) {
+	var risPopupURL = getXPathStr("href", doc, '//form[@id="pagepicker"]//a[contains(@href, "PrintRequest")][1]');
+    if (risPopupURL) {
+        // If page has RIS, use that
+	    var docParams = extractQueryValues(risPopupURL);
+	    var risURL = docParams.base 
+            + "CitationFile?kind=ris&handle=" + docParams.handle 
+            + "&div=" + docParams.div 
+            + "&id=" + docParams.id 
+            + "&base=js";
+	    var pdfPageURL = getXPathStr("href", doc, '//a[contains(@class, "updatediv")]');
+	    ZU.doGet(risURL, function(ris) {
+		    ris = fixRIS(ris);
+		    if (pdfPageURL) {
+			    ZU.doGet(pdfPageURL, function(pdfPage) {
+				    // Call to pdfPageURL prepares PDF for download via META refresh URL
+				    pdfURL = null;
+				    var m = pdfPage.match(/<META.*URL=([^\"]+)/);
+				    if (m) {
+					    var pdfURL = m[1];
+				    }
+				    translateRIS(ris, pdfURL);
+			    } , null , "UTF-8");
+		    } else {
+			    translateRIS(ris);
+		    }
+	    } , null , "UTF-8");
+    } else {
+        // No RIS available in page, try COinS
+        var COinS = getXPathStr("title", doc, '//span[contains(@class, "Z3988")]');
+        if (COinS) {
+            translateCOinS(COinS);
+        }
+    }
+}
+
+/*
+    *********
+    ** API **
+    *********
+*/
+
+function detectWeb (doc, url) {
+	if (url.indexOf("/LuceneSearch?") > -1) {
+		if (getSearchResults(doc)) {
+			return "multiple";
+		}
+	} else {
+		return "journalArticle";
+	}
+	return false;
+}
+
+function doWeb (doc, url) {
+	if (detectWeb(doc, url) === "multiple") {
+		Zotero.selectItems(getSearchResults(doc, url), function (items) {
+
+			if (!items) {
+				return true;
+			}
+			var urls = [];
+			for (var i in items) {
+				urls.push(i);
+			}
+
+			ZU.processDocuments(urls, scrapePage);
+		});
+	} else {
+		scrapePage(doc, url);
+	}
+}
+
+/** BEGIN TEST CASES **/
+var testCases = [
+	{
```

Comment:
Is it possible to add more tests for different cases above (multiples, two url patterns, ris, CoinS)